flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -131,6 +134,42 @@ void destroyPartitionRequestClient(ConnectionID connectionId, PartitionRequestCl
 		clients.remove(connectionId, client);
 	}
 
+	private NettyPartitionRequestClient connectChannelWithRetry(ConnectingChannel connectingChannel,
+										 ConnectionID connectionId, boolean needConnect)
+		throws IOException, InterruptedException {
+		int count = 0;
+		Exception exception = null;
+		do {
+			try {
+				if (needConnect) {
+					LOG.info("Connecting to {} at {} attempt", connectionId.getAddress(), count);
+					nettyClient.connect(connectionId.getAddress()).addListener(connectingChannel);
+				}
+
+				NettyPartitionRequestClient client = connectingChannel.waitForChannel();
+				clients.replace(connectionId, connectingChannel, client);
+				return client;
+			} catch (IOException | ChannelException e) {
+				LOG.error("Failed {} times to connect to {}", count, connectionId.getAddress(), e);
+				ConnectingChannel newConnectingChannel = new ConnectingChannel(connectionId, this);
+				clients.replace(connectionId, connectingChannel, newConnectingChannel);
```

Comment:
If the exception is thrown by `connect` and not `waitForChannel` then we use a new channel.
But other threads are waiting on the old channel - which will never be completed, right?

I guess this is the reason of the deadlock I see in Azure and locally.

Reply:
Yes. It is the reason of deadlock before adding the synchronized on connectionId change.